### PR TITLE
feat: Adding CEL validation for the `nodepool.spec.disruption`

### PR DESCRIPTION
**Pull request by engedaam**
Fixes #N/A 

**Description**
- Adding CEL validation in the effort of moving away from webhook validation.
- https://github.com/aws/karpenter-core/issues/103

**How was this change tested?**
- `make test`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/apis/crds/karpenter.sh_nodepools.yaml [hunk 1]
```diff
@@ -78,7 +81,15 @@ spec:
                       node upgrade, memory leak protection, and disruption testing.
                     pattern: ^(([0-9]+(s|m|h))+)|(Never)$
                     type: string
+                    x-kubernetes-validations:
+                    - message: expireAfter cannot be negative
+                      rule: '!self.startsWith(''-'')'
                 type: object
+                x-kubernetes-validations:
+                - rule: 'has(self.consolidateAfter) ? self.consolidationPolicy!=''WhenUnderutilized''
```
Comment (jonathan-innis) [inline pkg/apis/crds/karpenter.sh_nodepools.yaml:89]: Just curious. Can you do spaces in between `!=` and `==` or do you have to have no spaces in there

Comment (jonathan-innis) [inline pkg/apis/crds/karpenter.sh_nodepools.yaml:89]: Should we add a message if this validation fails?

File: pkg/controllers/nodeclaim/disruption/nodeclaim_test.go [hunk 2]
```diff
@@ -63,7 +63,7 @@ var _ = Describe("NodeClaim/Disruption", func() {
 	})
 	It("should remove multiple disruption conditions simultaneously", func() {
 		nodePool.Spec.Disruption.ExpireAfter.Duration = nil
-		nodePool.Spec.Disruption.ConsolidateAfter = &v1beta1.NillableDuration{Duration: nil}
```
Comment (jonathan-innis) [inline pkg/controllers/nodeclaim/disruption/nodeclaim_test.go:66]: Does this not succeed now? We should allow this to pass

File: pkg/apis/v1beta1/nodepool_validation_cel_test.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/aws/karpenter-core/pkg/apis"
+	. "github.com/aws/karpenter-core/pkg/apis/v1beta1"
+	"github.com/aws/karpenter-core/pkg/operator/scheme"
+	"github.com/aws/karpenter-core/pkg/test"
+)
+
+var _ = Describe("CEL/Validation", func() {
```
Comment (jonathan-innis) [inline pkg/apis/v1beta1/nodepool_validation_cel_test.go:34]: Does it make sense to throw this into a separate doc or to put this into the same validation suite_test?

Comment (jonathan-innis) [reply pkg/apis/v1beta1/nodepool_validation_cel_test.go:34]: I could go either way on this. Just curious your thoughts.

Comment (engedaam) [reply pkg/apis/v1beta1/nodepool_validation_cel_test.go:34]: I was thinking of following the same ideas that was used in the nodeClaim testing. Also, I thought the file would be really long, and would be good to separate it out

Comment (jonathan-innis) [reply pkg/apis/v1beta1/nodepool_validation_cel_test.go:34]: I'm good either way. Makes sense to me 👍

File: pkg/apis/crds/karpenter.sh_nodepools.yaml [hunk 4]
```diff
@@ -78,7 +81,18 @@ spec:
                       node upgrade, memory leak protection, and disruption testing.
                     pattern: ^(([0-9]+(s|m|h))+)|(Never)$
                     type: string
+                    x-kubernetes-validations:
+                    - message: expireAfter cannot be negative
```
Comment (jonathan-innis) [inline pkg/apis/crds/karpenter.sh_nodepools.yaml:85]: This is already enforced by the regex so I don't think that we need this anymore

File: pkg/apis/v1beta1/nodepool_validation_cel_test.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+			nodePool.Spec.Disruption.ConsolidateAfter = &NillableDuration{Duration: lo.ToPtr(lo.Must(time.ParseDuration("30s")))}
+			nodePool.Spec.Disruption.ConsolidationPolicy = ConsolidationPolicyWhenEmpty
+			Expect(env.Client.Create(ctx, nodePool)).To(Succeed())
+		})
+		It("should succeed when setting consolidateAfter with consolidationPolicy=WhenEmpty", func() {
+			nodePool.Spec.Disruption.ConsolidateAfter = &NillableDuration{Duration: lo.ToPtr(lo.Must(time.ParseDuration("30s")))}
+			nodePool.Spec.Disruption.ConsolidationPolicy = ConsolidationPolicyWhenEmpty
+			Expect(env.Client.Create(ctx, nodePool)).To(Succeed())
+		})
+		It("should fail when setting consolidateAfter with consolidationPolicy=WhenUnderutilized", func() {
```
Comment (jonathan-innis) [inline pkg/apis/v1beta1/nodepool_validation_cel_test.go:91]: Can we also have a check that validates that `consolidationPolicy=WhenUnderutilized` works with `consolidateAfter=Never`

File: pkg/apis/v1beta1/nodepool_validation_cel_test.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+			nodePool.Spec.Disruption.ConsolidateAfter = &NillableDuration{Duration: lo.ToPtr(lo.Must(time.ParseDuration("30s")))}
+			nodePool.Spec.Disruption.ConsolidationPolicy = ConsolidationPolicyWhenUnderutilized
+			Expect(env.Client.Create(ctx, nodePool)).ToNot(Succeed())
+		})
+		It("should succeed when not setting consolidateAfter with consolidationPolicy=WhenUnderutilized", func() {
```
Comment (jonathan-innis) [inline pkg/apis/v1beta1/nodepool_validation_cel_test.go:96]:
```suggestion
		It("should succeed when not setting consolidateAfter to 'Never' with consolidationPolicy=WhenUnderutilized", func() {
```

**Review by jonathan-innis — APPROVED** [review]
Nice work 🎉

**Pull request merged**